<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto</font></h1>


### Assignment for Week 3

For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.

Author : Muhammad Adisatriyo Pratama

References : [Coursera Applied Data Science Capstone](https://www.coursera.org/learn/applied-data-science-capstone)

December 2020

Before we get the data and start exploring it, let's download all the dependencies that we will need.


In [1]:
!pip install folium

In [2]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

# !pip install -c folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## Problem 1

Importing data from Wikipedia : [Postal Codes of Canada](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M)

In [3]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

# make a requests
wiki_url = requests.get(url)
wiki_url

<Response [200]>

wiki_url returns Response 200 wich means connection is established

In [4]:
# Read html from wiki_url using pandas library
html_wiki = pd.read_html(wiki_url.text)

# check type
print(f'Type of html_wiki : {type(html_wiki)}')

# length of html_wiki
print(f'Length of html_wiki : {len(html_wiki)}')

Type of html_wiki : <class 'list'>
Length of html_wiki : 3


In [5]:
# Check first few data from list
html_wiki[0:5]

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 5           M6A        North York   
 6           M7A  Downtown Toronto   
 7           M8A      Not assigned   
 8           M9A         Etobicoke   
 9           M1B       Scarborough   
 10          M2B      Not assigned   
 11          M3B        North York   
 12          M4B         East York   
 13          M5B  Downtown Toronto   
 14          M6B        North York   
 15          M7B      Not assigned   
 16          M8B      Not assigned   
 17          M9B         Etobicoke   
 18          M1C       Scarborough   
 19          M2C      Not assigned   
 20          M3C        North York   
 21          M4C         East York   
 22          M5C  Downtown Toronto   
 23          M6C              York   
 24          M7C      Not assigned   
 25         

We want first table wich is consist of three columns: PostalCode, Borough, and Neighborhood.

In [10]:
# We want first table on the list
df = html_wiki[0]
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

 Ignore cells with a borough that is **Not assigned**.

In [11]:
# Dropping 'Not assigned' borough
df[df['Borough'] == 'Not assigned']


Postal Code       Borough Neighbourhood
0           M1A  Not assigned  Not assigned
1           M2A  Not assigned  Not assigned
7           M8A  Not assigned  Not assigned
10          M2B  Not assigned  Not assigned
15          M7B  Not assigned  Not assigned
16          M8B  Not assigned  Not assigned
19          M2C  Not assigned  Not assigned
24          M7C  Not assigned  Not assigned
25          M8C  Not assigned  Not assigned
28          M2E  Not assigned  Not assigned
29          M3E  Not assigned  Not assigned
33          M7E  Not assigned  Not assigned
34          M8E  Not assigned  Not assigned
35          M9E  Not assigned  Not assigned
37          M2G  Not assigned  Not assigned
38          M3G  Not assigned  Not assigned
42          M7G  Not assigned  Not assigned
43          M8G  Not assigned  Not assigned
44          M9G  Not assigned  Not assigned
51          M7H  Not assigned  Not assigned
52          M8H  Not assigned  Not assigned
53          M9H  Not assigned  Not assigned
60          M7J  Not assigned  Not assigned
61          M8J  Not assigned  Not assigned
62          M9J  Not assigned  Not assigned
69          M7K  Not assigned  Not assigned
70          M8K  Not assigned  Not assigned
71          M9K  Not assigned  Not assigned
78          M7L  Not assigned  Not assigned
79          M8L  Not assigned  Not assigned
87          M7M  Not assigned  Not assigned
88          M8M  Not assigned  Not assigned
96          M7N  Not assigned  Not assigned
97          M8N  Not assigned  Not assigned
101         M3P  Not assigned  Not assigned
105         M7P  Not assigned  Not assigned
106         M8P  Not assigned  Not assigned
110         M3R  Not assigned  Not assigned
115         M8R  Not assigned  Not assigned
118         M2S  Not assigned  Not assigned
119         M3S  Not assigned  Not assigned
123         M7S  Not assigned  Not assigned
124         M8S  Not assigned  Not assigned
125         M9S  Not assigned  Not assigned
127         M2T  Not assigned  Not assigned
128         M3T  Not assigned  Not assigned
131         M6T  Not assigned  Not assigned
132         M7T  Not assigned  Not assigned
133         M8T  Not assigned  Not assigned
134         M9T  Not assigned  Not assigned
136         M2V  Not assigned  Not assigned
137         M3V  Not assigned  Not assigned
140         M6V  Not assigned  Not assigned
141         M7V  Not assigned  Not assigned
145         M2W  Not assigned  Not assigned
146         M3W  Not assigned  Not assigned
149         M6W  Not assigned  Not assigned
150         M7W  Not assigned  Not assigned
154         M2X  Not assigned  Not assigned
155         M3X  Not assigned  Not assigned
158         M6X  Not assigned  Not assigned
159         M7X  Not assigned  Not assigned
161         M9X  Not assigned  Not assigned
162         M1Y  Not assigned  Not assigned
163         M2Y  Not assigned  Not assigned
164         M3Y  Not assigned  Not assigned
166         M5Y  Not assigned  Not assigned
167         M6Y  Not assigned  Not assigned
170         M9Y  Not assigned  Not assigned
171         M1Z  Not assigned  Not assigned
172         M2Z  Not assigned  Not assigned
173         M3Z  Not assigned  Not assigned
174         M4Z  Not assigned  Not assigned
175         M5Z  Not assigned  Not assigned
176         M6Z  Not assigned  Not assigned
177         M7Z  Not assigned  Not assigned
179         M9Z  Not assigned  Not assigned

In [13]:
df = df[df['Borough'] != 'Not assigned']
df.head(10)

Postal Code           Borough                                Neighbourhood
2          M3A        North York                                    Parkwoods
3          M4A        North York                             Victoria Village
4          M5A  Downtown Toronto                    Regent Park, Harbourfront
5          M6A        North York             Lawrence Manor, Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
8          M9A         Etobicoke      Islington Avenue, Humber Valley Village
9          M1B       Scarborough                               Malvern, Rouge
11         M3B        North York                                    Don Mills
12         M4B         East York              Parkview Hill, Woodbine Gardens
13         M5B  Downtown Toronto                     Garden District, Ryerson

More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

So we will group the records based on Postal Code

In [23]:
# Group by 'Postal Code'
clean_df = df.groupby(['Postal Code']).head()
clean_df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [24]:
# Check 'Not assigned' Neighbourhood
clean_df[clean_df['Neighbourhood'] == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

There is no  *'Not assigned'*  Neighbourhood left

#### Resetting index for clean_df to start from 0

In [25]:
# resetting index for clean_df to start from 0
clean_df = clean_df.reset_index(drop=True)
clean_df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Vic

### Shape of clean_df containing information about Neighbourhood in Toronto

In [26]:
clean_df.shape

(103, 3)

clean_df contains 103 rows and 3 columns

## Problem 2
### Discovering latitude and longitude

We were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code.

Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

Use the Geocoder package or the csv file to add latitude and longitude to the the dataframe:

In [27]:
# Installing geocoder
!pip install geocoder

     |████████████████████████████████| 98 kB 5.4 MB/s eta 0:00:011


In [28]:
# Import geocoder
import geocoder

Given link to a csv file that has the geographical coorinates of each postal code : http://cocl.us/Geospatial_data

So we read that csv using pandas and assign it to geo_data

In [29]:
geo_data = pd.read_csv('http://cocl.us/Geospatial_data')
geo_data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

Check the length of neighborhood data (**clean_df**) and geospatial data (**geo_data**) 

In [31]:
print(f'Length of Neighborhood data : {len(clean_df)}')
print(f'Length of Geospatial data : {len(geo_data)}')

Length of Neighborhood data : 103
Length of Geospatial data : 103


Check data types for each columns in neighborhood data and geospatial data

In [33]:
print(f'Data types of Neighborhood data : \n{clean_df.dtypes}\n')
print(f'Data types of Geospatial data : \n{geo_data.dtypes}')

Data types of Neighborhood data : 
Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

Data types of Geospatial data : 
Postal Code     object
Latitude       float64
Longitude      float64
dtype: object


Join both dataframe on 'Postal Code' column

In [35]:
geo_data.set_index('Postal Code')

Latitude  Longitude
Postal Code                      
M1B          43.806686 -79.194353
M1C          43.784535 -79.160497
M1E          43.763573 -79.188711
M1G          43.770992 -79.216917
M1H          43.773136 -79.239476
M1J          43.744734 -79.239476
M1K          43.727929 -79.262029
M1L          43.711112 -79.284577
M1M          43.716316 -79.239476
M1N          43.692657 -79.264848
M1P          43.757410 -79.273304
M1R          43.750072 -79.295849
M1S          43.794200 -79.262029
M1T          43.781638 -79.304302
M1V          43.815252 -79.284577
M1W          43.799525 -79.318389
M1X          43.836125 -79.205636
M2H          43.803762 -79.363452
M2J          43.778517 -79.346556
M2K          43.786947 -79.385975
M2L          43.757490 -79.374714
M2M          43.789053 -79.408493
M2N          43.770120 -79.408493
M2P          43.752758 -79.400049
M2R          43.782736 -79.442259
M3A          43.753259 -79.329656
M3B          43.745906 -79.352188
M3C          43.725900 -79.340923
M3H          43.754328 -79.442259
M3J          43.767980 -79.487262
M3K          43.737473 -79.464763
M3L          43.739015 -79.506944
M3M          43.728496 -79.495697
M3N          43.761631 -79.520999
M4A          43.725882 -79.315572
M4B          43.706397 -79.309937
M4C          43.695344 -79.318389
M4E          43.676357 -79.293031
M4G          43.709060 -79.363452
M4H          43.705369 -79.349372
M4J          43.685347 -79.338106
M4K          43.679557 -79.352188
M4L          43.668999 -79.315572
M4M          43.659526 -79.340923
M4N          43.728020 -79.388790
M4P          43.712751 -79.390197
M4R          43.715383 -79.405678
M4S          43.704324 -79.388790
M4T          43.689574 -79.383160
M4V          43.686412 -79.400049
M4W          43.679563 -79.377529
M4X          43.667967 -79.367675
M4Y          43.665860 -79.383160
M5A          43.654260 -79.360636
M5B          43.657162 -79.378937
M5C          43.651494 -79.375418
M5E          43.644771 -79.373306
M5G          43.657952 -79.387383
M5H          43.650571 -79.384568
M5J          43.640816 -79.381752
M5K          43.647177 -79.381576
M5L          43.648198 -79.379817
M5M          43.733283 -79.419750
M5N          43.711695 -79.416936
M5P          43.696948 -79.411307
M5R          43.672710 -79.405678
M5S          43.662696 -79.400049
M5T          43.653206 -79.400049
M5V          43.628947 -79.394420
M5W          43.646435 -79.374846
M5X          43.648429 -79.382280
M6A          43.718518 -79.464763
M6B          43.709577 -79.445073
M6C          43.693781 -79.428191
M6E          43.689026 -79.453512
M6G          43.669542 -79.422564
M6H          43.669005 -79.442259
M6J          43.647927 -79.419750
M6K          43.636847 -79.428191
M6L          43.713756 -79.490074
M6M          43.691116 -79.476013
M6N          43.673185 -79.487262
M6P          43.661608 -79.464763
M6R          43.648960 -79.456325
M6S          43.651571 -79.484450
M7A          43.662301 -79.389494
M7R          43.636966 -79.615819
M7Y          43.662744 -79.321558
M8V          43.605647 -79.501321
M8W          43.602414 -79.543484
M8X          43.653654 -79.506944
M8Y          43.636258 -79.498509
M8Z          43.628841 -79.520999
M9A          43.667856 -79.532242
M9B          43.650943 -79.554724
M9C          43.643515 -79.577201
M9L          43.756303 -79.565963
M9M          43.724766 -79.532242
M9N          43.706876 -79.518188
M9P          43.696319 -79.532242
M9R          43.688905 -79.554724
M9V          43.739416 -79.588437
M9W          43.706748 -79.594054

In [36]:
neighborhood_df = clean_df.join(geo_data.set_index('Postal Code'), on ='Postal Code', how='inner')
neighborhood_df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

Shape of neighborhood_df

In [37]:
neighborhood_df.shape

(103, 5)

In neighborhood_df we got 103 rows and now we have 5 columns adding latitude and longitude to the dataframe

## Problem 3

### Explore and cluster the neighborhoods in Toronto


- Get coordinate of Toronto using geopy.geocoders.Nominatim object
- Visualize map of Toronto with Folium library
- Get neighborhoods data using Foursquare API
- Find similarities of categories uising K-Means Clustering

In [38]:
# Find coordinates of Toronto
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Visualizing map of Toronto

In [39]:
toronto_map = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(neighborhood_df['Latitude'], neighborhood_df['Longitude'], neighborhood_df['Borough'], neighborhood_df['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(toronto_map)
    
toronto_map

Defining Foursquare API Credentials and Version

In [41]:
# @hidden cells
CLIENT_ID = 'L00XEGIHDO1OSWE2JBEV5WPRY4IILAHBLKKZ54WGRU51OTPU' # your Foursquare ID
CLIENT_SECRET = 'VZ2VJZJIZA1XAY4OXXOTMQHVMFUPVSPTBI21LUJSIDWE1K4U' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: L00XEGIHDO1OSWE2JBEV5WPRY4IILAHBLKKZ54WGRU51OTPU
CLIENT_SECRET:VZ2VJZJIZA1XAY4OXXOTMQHVMFUPVSPTBI21LUJSIDWE1K4U


Get a data about venues in Toronto using Foursquare API

In [49]:
# Function that return latitude, longitude, venues, and venue_categories in neighborhood_df
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    # create an empty list
    venues_list=[]
    
    # for loop that iterate through dataframe
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

        
    # Create pandas dataframe from venues_list
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

#### Get venues data for each neighborhood in Toronto

In [50]:
toronto_venues = getNearbyVenues(neighborhood_df['Neighbourhood'], neighborhood_df['Latitude'], neighborhood_df['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

#### Check the size of the resulting dataframe

In [51]:
print(f'Shape of toronto_venues dataframe : {toronto_venues.shape}\n')
print('Head of toronto_venues dataframe : ')
toronto_venues.head()

Shape of toronto_venues dataframe : (1322, 5)

Head of toronto_venues dataframe : 


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3             Tim Hortons            Coffee Shop  
4               Portugril  Portuguese Restaurant

#### Check how many venues were returned for each neighborhood

In [52]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               5   
Alderwood, Long Branch                                                  7   
Bathurst Manor, Wilson Heights, Downsview North                        21   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      22   
Berczy Park                                                            30   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  13   
Cedarbrae                                                               8   
Central Bay Street                                                     30   
Christie                                                               16   
Church and Wellesley                                                   30   
Clarks Corners, Tam O'Shanter, Sullivan                                12   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                         30   
Davisville                                                             30   
Davisville North                                                        9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        5   
Don Mills                                                              23   
Dorset Park, Wexford Heights, Scarborough Town ...                      5   
Downsview                                                              16   
Dufferin, Dovercourt Village                                           13   
East Toronto, Broadview North (Old East York)                           3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      8   
Fairview, Henry Farm, Oriole                                           30   
First Canadian Place, Underground city                                 30   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                               30   
Glencairn                                                               5   
Golden Mile, Clairlea, Oakridge                                        10   
Guildwood, Morningside, West Hill                                       8   
Harbourfront East, Union Station, Toronto Islands                      30   
High Park, The Junction South                                          25   
Hillcrest Village                                                       5   
Humber Summit                                                           2   
Humberlea, Emery                                                        1   
Humewood-Cedarvale                                                      5   
India Bazaar, The Beaches West                                         19   
Kennedy Park, Ionview, East Birchmount Park                             4   
Kensington Market, Chinatown, Grange Park                              30   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                                       13   
Lawrence Park                                                           3   
Leaside                                                                30   
Little Portugal, Trinity                                               30   
Malvern, Rouge                                                          1   
Milliken, Agincourt North, Steeles East, L'Amor.

#### How many unique categories can be curated from all the returned venues

In [53]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 uniques categories.


### Using One Hot Encoding to create 237 categories as features/columns

In [54]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Basketball Stadium  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                   0         0           0                   0          0   
1                   0         0           0                   0          0   
2                   0         0           0                   0          0   
3                   0         0           0                   0          0   
4                   0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Breakfast Spot  Brewery  \
0       0              0          0         0               0        0   
1       0              0          0         0               0        0   
2       0              0          0         0               0        0   
3       0              0          0         0               0        0   
4       0              0          0         0               0        0   

   Bridal Shop  Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  \
0            0                0             0              0         0   
1            0                0             0              0         0   
2            0                0             0              0         0   
3            0                0             0              0         0   
4            0                0             0              0         0   

   Bus Station  Business Service  Butcher  Café  Cajun / Creole Restaurant  \
0            0                 0        0     0                          0   
1            0                 0        0     0                          0   
2            0                 0        0     0                          0   
3            0                 0        0     0            

In [55]:
# Shape of toronto_onehot
toronto_onehot.shape

(1322, 237)

#### Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [56]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                           Agincourt     0.000000   
1                              Alderwood, Long Branch     0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North     0.000000   
3                                     Bayview Village     0.000000   
4                   Bedford Park, Lawrence Manor East     0.000000   
5                                         Berczy Park     0.000000   
6                         Birch Cliff, Cliffside West     0.000000   
7        Brockton, Parkdale Village, Exhibition Place     0.000000   
8   Business reply mail Processing Centre, South C...     0.000000   
9   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
10                                Caledonia-Fairbanks     0.000000   
11              Canada Post Gateway Processing Centre     0.000000   
12                                          Cedarbrae     0.000000   
13                                 Central Bay Street     0.033333   
14                                           Christie     0.000000   
15                               Church and Wellesley     0.000000   
16            Clarks Corners, Tam O'Shanter, Sullivan     0.000000   
17    Cliffside, Cliffcrest, Scarborough Village West     0.000000   
18                     Commerce Court, Victoria Hotel     0.000000   
19                                         Davisville     0.000000   
20                                   Davisville North     0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn     0.000000   
22                                          Don Mills     0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...     0.000000   
24                                          Downsview     0.000000   
25                       Dufferin, Dovercourt Village     0.000000   
26      East Toronto, Broadview North (Old East York)     0.000000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...     0.000000   
28                       Fairview, Henry Farm, Oriole     0.000000   
29             First Canadian Place, Underground city     0.000000   
30    Forest Hill North & West, Forest Hill Road Park     0.000000   
31                           Garden District, Ryerson     0.000000   
32                                          Glencairn     0.000000   
33                    Golden Mile, Clairlea, Oakridge     0.000000   
34                  Guildwood, Morningside, West Hill     0.000000   
35  Harbourfront East, Union Station, Toronto Islands     0.000000   
36                      High Park, The Junction South     0.000000   
37                                  Hillcrest Village     0.000000   
38                                      Humber Summit     0.000000   
39                                   Humberlea, Emery     0.000000   
40                                 Humewood-Cedarvale     0.000000   
41                     India Bazaar, The Beaches West     0.000000   
42        Kennedy Park, Ionview, East Birchmount Park     0.000000   
43          Kensington Market, Chinatown, Grange Park     0.000000   
44  Kingsview Village, St. Phillips, Martin Grove ...     0.000000   
45                   Lawrence Manor, Lawrence Heights     0.000000   
46                                      Lawrence Park     0.000000   
47                                            Leaside     0.000000   
48                           Little Portugal, Trinity     0.033333   
49                                     Malvern, Rouge     0.000000   
50  Milliken, Agincourt North, Steeles East, L'Amo...     0.000000   
51  Mimico NW, The Queensway West, South of Bloor,...     0.000000   
52                        Moore Park, Summerhill East     0.000000   
53       New Toronto, Mimico South, Humber Bay Shores     0.000000   
54           North Park, Maple Leaf Park, Upwood Park     0.000000   
55                  North Toronto West, Lawrence Park     0.055556   
56                Northwest, West Humber - Clairvil

In [57]:
# Shape of toronto_grouped
toronto_grouped.shape

(96, 237)

#### Create a _pandas_ dataframe for each neighborhood with the top 10 most common venues

In [59]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [70]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village                       Café   
4                Bedford Park, Lawrence Manor East                Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge          Skating Rink        Clothing Store   
1              Pharmacy                   Gym           Coffee Shop   
2                  Bank         Shopping Mall           Pizza Place   
3                  Bank    Chinese Restaurant   Japanese Restaurant   
4        Sandwich Place    Italian Restaurant      Greek Restaurant   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot                Women's Store         Deli / Bodega   
1        Sandwich Place                          Pub          Dessert Shop   
2      Sushi Restaurant               Ice Cream Shop            Restaurant   
3         Deli / Bodega  Eastern European Restaurant   Dumpling Restaurant   
4             Juice Bar                         Café      Sushi Restaurant   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0             Drugstore                 Donut Shop                Dog Run  
1          Dance Studio              Deli / Bodega       Department Store  
2         Deli / Bodega  Middle Eastern Restaurant     Chinese Restaurant  
3             Drugstore                 Donut Shop                Dog Run  
4     Indian Restaurant                        Pub             Restaurant

In [71]:
neighborhoods_venues_sorted

Neighborhood  \
0                                           Agincourt   
1                              Alderwood, Long Branch   
2     Bathurst Manor, Wilson Heights, Downsview North   
3                                     Bayview Village   
4                   Bedford Park, Lawrence Manor East   
5                                         Berczy Park   
6                         Birch Cliff, Cliffside West   
7        Brockton, Parkdale Village, Exhibition Place   
8   Business reply mail Processing Centre, South C...   
9   CN Tower, King and Spadina, Railway Lands, Har...   
10                                Caledonia-Fairbanks   
11              Canada Post Gateway Processing Centre   
12                                          Cedarbrae   
13                                 Central Bay Street   
14                                           Christie   
15                               Church and Wellesley   
16            Clarks Corners, Tam O'Shanter, Sullivan   
17    Cliffside, Cliffcrest, Scarborough Village West   
18                     Commerce Court, Victoria Hotel   
19                                         Davisville   
20                                   Davisville North   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn   
22                                          Don Mills   
23  Dorset Park, Wexford Heights, Scarborough Town...   
24                                          Downsview   
25                       Dufferin, Dovercourt Village   
26      East Toronto, Broadview North (Old East York)   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...   
28                       Fairview, Henry Farm, Oriole   
29             First Canadian Place, Underground city   
30    Forest Hill North & West, Forest Hill Road Park   
31                           Garden District, Ryerson   
32                                          Glencairn   
33                    Golden Mile, Clairlea, Oakridge   
34                  Guildwood, Morningside, West Hill   
35  Harbourfront East, Union Station, Toronto Islands   
36                      High Park, The Junction South   
37                                  Hillcrest Village   
38                                      Humber Summit   
39                                   Humberlea, Emery   
40                                 Humewood-Cedarvale   
41                     India Bazaar, The Beaches West   
42        Kennedy Park, Ionview, East Birchmount Park   
43          Kensington Market, Chinatown, Grange Park   
44  Kingsview Village, St. Phillips, Martin Grove ...   
45                   Lawrence Manor, Lawrence Heights   
46                                      Lawrence Park   
47                                            Leaside   
48                           Little Portugal, Trinity   
49                                     Malvern, Rouge   
50  Milliken, Agincourt North, Steeles East, L'Amo...   
51  Mimico NW, The Queensway West, South of Bloor,...   
52                        Moore Park, Summerhill East   
53       New Toronto, Mimico South, Humber Bay Shores   
54           North Park, Maple Leaf Park, Upwood Park   
55                  North Toronto West, Lawrence Park   
56                Northwest, West Humber - Clairville   
57                    Northwood Park, York University   
58  Old Mill South, King's Mill Park, Sunnylea, Hu...   
59                             Parkdale, Roncesvalles   
60                    Parkview Hill, Woodbine Gardens   
61                                          Parkwoods   
62        Queen's Park, Ontario Provincial Government   
63                          Regent Park, Harbourfront   
64                           Richmond, Adelaide, King   
65                                           Rosedale   
66                                           Roselawn   
67             Rouge Hill, Port Union, Highland Creek   
68                                 Runnymede, Swansea   
69                      Runnymede, The Junction Nor

In [72]:
# Shape of neighborhoods_venues_sorted
neighborhoods_venues_sorted.shape

(96, 11)

# Clustering Neighborhood

Run _k_-means to cluster the neighborhood into 5 clusters.


In [73]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [74]:
# Renaming Neighborhoud columns from neighborhood_df to Neighborhood for joining dataframe
neighborhood_df.rename(columns={'Neighbourhood': 'Neighborhood'}, inplace=True)
neighborhood_df.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [75]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neighborhood_df

# merge toronto_grouped with neighborhood_df to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0          Intersection   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop           Women's Store          Dance Studio   
1           Pizza Place             Coffee Shop     French Restaurant   
2                  Park                  Bakery                   Pub   
3             Gift Shop  Furniture / Home Store           Event Space   
4    College Auditorium                     Bar              Beer Bar   

  5th Most Common Venue  6th Most Common Venue  7th Most Common Venue  \
0   Dumpling Restaurant              Drugstore             Donut Shop   
1          Hockey Arena  Portuguese Restaurant     Dim Sum Restaurant   
2        Breakfast Spot            Yoga Studio      French Restaurant   
3           Coffee Shop          Women's Store  Vietnamese Restaurant   
4         Smoothie Shop         Sandwich Place                   Café   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Dog Run   Distribution Center         Discount Store  
1          Deli / Bodega      Department Store           Dessert Shop  
2           Dessert Shop                   Spa            Event Space  
3               Boutique     Accessories Store              Drugstore  
4  Portuguese Restaurant    Chinese Restaurant                   Park

In [81]:
# Check na value for 'Cluster Label'
toronto_merged['Cluster Labels'].isna().sum()

3

In [82]:
# Drop NA
toronto_merged_clean = toronto_merged.dropna(subset=['Cluster Labels'])

In [83]:
# New shape of toronto_merged_clean
toronto_merged_clean.shape

(100, 16)

## Visualize the resulting clusters

In [86]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_clean['Latitude'], toronto_merged_clean['Longitude'], toronto_merged_clean['Neighborhood'], toronto_merged_clean['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine each Clusters

#### Cluster 1


In [87]:
toronto_merged_clean.loc[toronto_merged_clean['Cluster Labels'] == 0, toronto_merged_clean.columns[[1] + list(range(5, toronto_merged_clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
35         East York             0.0                  Park   
49        North York             0.0                  Park   
61   Central Toronto             0.0                  Park   
64              York             0.0                  Park   
66        North York             0.0                  Park   
85       Scarborough             0.0                  Park   
91  Downtown Toronto             0.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Food & Drink Shop         Women's Store          Dance Studio   
21               Women's Store                  Pool           Curling Ice   
35           Convenience Store          Intersection         Women's Store   
49  Construction & Landscaping                Bakery         Women's Store   
61                 Swim School              Bus Line         Women's Store   
64               Women's Store           Curling Ice   Dumpling Restaurant   
66           Convenience Store         Women's Store          Dance Studio   
85                  Playground                Bakery          Intersection   
91                  Playground                 Trail         Women's Store   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0    Dumpling Restaurant                    Drugstore            Donut Shop   
21   Dumpling Restaurant                    Drugstore            Donut Shop   
35         Deli / Bodega  Eastern European Restaurant   Dumpling Restaurant   
49         Deli / Bodega  Eastern European Restaurant   Dumpling Restaurant   
61         Deli / Bodega          Dumpling Restaurant             Drugstore   
64             Drugstore                   Donut Shop               Dog Run   
66   Dumpling Restaurant                    Drugstore            Donut Shop   
85         Women's Store                Deli / Bodega   Dumpling Restaurant   
91           Curling Ice                    Drugstore            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Dog Run   Distribution Center         Discount Store  
21               Dog Run   Distribution Center         Discount Store  
35             Drugstore            Donut Shop                Dog Run  
49             Drugstore            Donut Shop                Dog Run  
61            Donut Shop               Dog Run    Distribution Center  
64   Distribution Center        Discount Store                  Diner  
66               Dog Run   Distribution Center         Discount Store  
85             Drugstore            Donut Shop                Dog Run  
91               Dog Run   Distribution Center         Discount Store

#### Cluster 2


In [88]:
toronto_merged_clean.loc[toronto_merged_clean['Cluster Labels'] == 1, toronto_merged_clean.columns[[1] + list(range(5, toronto_merged_clean.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
1          North York             1.0               Intersection   
2    Downtown Toronto             1.0                Coffee Shop   
3          North York             1.0             Clothing Store   
4    Downtown Toronto             1.0                Coffee Shop   
7          North York             1.0                        Gym   
8           East York             1.0                Pizza Place   
9    Downtown Toronto             1.0                       Café   
10         North York             1.0                       Park   
12        Scarborough             1.0                        Bar   
13         North York             1.0                        Gym   
14          East York             1.0               Skating Rink   
15   Downtown Toronto             1.0                  Gastropub   
16               York             1.0               Tennis Court   
17          Etobicoke             1.0                   Pharmacy   
18        Scarborough             1.0                 Restaurant   
19       East Toronto             1.0          Health Food Store   
20   Downtown Toronto             1.0             Farmers Market   
22        Scarborough             1.0                Coffee Shop   
23          East York             1.0                Coffee Shop   
24   Downtown Toronto             1.0                Coffee Shop   
25   Downtown Toronto             1.0              Grocery Store   
26        Scarborough             1.0         Athletics & Sports   
27         North York             1.0                Golf Course   
28         North York             1.0                Coffee Shop   
29          East York             1.0             Sandwich Place   
30   Downtown Toronto             1.0                Coffee Shop   
31       West Toronto             1.0                   Pharmacy   
32        Scarborough             1.0                 Playground   
33         North York             1.0             Clothing Store   
34         North York             1.0                        Bar   
36   Downtown Toronto             1.0                       Park   
37       West Toronto             1.0                        Bar   
38        Scarborough             1.0              Train Station   
39         North York             1.0                       Café   
40         North York             1.0              Grocery Store   
41       East Toronto             1.0           Greek Restaurant   
42   Downtown Toronto             1.0                Coffee Shop   
43       West Toronto             1.0                       Café   
44        Scarborough             1.0                   Bus Line   
46         North York             1.0              Grocery Store   
47       East Toronto             1.0                       Park   
48   Downtown Toronto             1.0                       Café   
50         North York             1.0                Pizza Place   
51        Scarborough             1.0        American Restaurant   
53         North York             1.0              Grocery Store   
54       East Toronto             1.0                Coffee Shop   
55         North York             1.0                Coffee Shop   
56               York             1.0                 Restaurant   
57         North York             1.0             Baseball Field   
58        Scarborough             1.0                       Café   
59         North York             1.0           Ramen Restaurant   
60         North York             1.0              Grocery Store   
62    Central Toronto             1.0                     Garden   
63               York             1.0             Breakfast Spot   
65        Scarborough             1.0          Indian Restaurant   
67    Central Toronto             1.0               Dance Studio   
68    Central Toronto             1.0                       Park   
69       West Toronto             1.0         Mexican Restaurant   
70

#### Cluster 3

In [89]:
toronto_merged_clean.loc[toronto_merged_clean['Cluster Labels'] == 2, toronto_merged_clean.columns[[1] + list(range(5, toronto_merged_clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York             2.0   Martial Arts School         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45          Dance Studio   Dumpling Restaurant             Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45            Donut Shop               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
45        Discount Store                  Diner

#### Cluster 4

In [90]:
toronto_merged_clean.loc[toronto_merged_clean['Cluster Labels'] == 3, toronto_merged_clean.columns[[1] + list(range(5, toronto_merged_clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             3.0            Print Shop         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11           Curling Ice   Dumpling Restaurant             Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11            Donut Shop               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
11        Discount Store                  Diner

#### Cluster 5

In [91]:
toronto_merged_clean.loc[toronto_merged_clean['Cluster Labels'] == 4, toronto_merged_clean.columns[[1] + list(range(5, toronto_merged_clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             4.0  Fast Food Restaurant         Women's Store   

  3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
6           Escape Room  Eastern European Restaurant   Dumpling Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6             Drugstore            Donut Shop               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
6   Distribution Center         Discount Store

## Finally finished Clustering Toronto with 5 Clusters using K-Means Clustering

## Author : Muhammad Adisatriyo Pratama
## References : [Coursera Applied Data Science Capstone](https://www.coursera.org/learn/applied-data-science-capstone)

### December 2020